defaults.go

```diff
@@ -253,7 +253,18 @@ func IsMsg(buf []byte) error {
 
 // IsFqdn checks if a domain name is fully qualified.
 func IsFqdn(s string) bool {
-	return strings.HasSuffix(s, ".")
+	s2 := strings.TrimSuffix(s, ".")
+	if s == s2 {
+		return false
+	}
+
+	i := strings.LastIndexFunc(s2, func(r rune) bool {
+		return r != '\\'
+	})
+
+	// Test whether we have an even number of escape sequences before
+	// the dot or none.
+	return (len(s2)-i)%2 != 0
```

Comment:
I'm not 100% sure I've got the logic right here, but it passes every test I can throw at it.

Reply:
@miekg That's exactly what this does and there's no allocation. `TrimSuffix` will remove the dot if it exists by slicing `s`, `s == s2` checks if there was a dot, `strings.LastIndexFunc` finds the last non-backslash in the name, and then `(len(s2)-i)%2` calculates if it's an even or odd number of backslashes.